feat(cli): initial version of `subo doc` to build and test docs snippets

denopink opened the pull request:
# Overview

PR for #22 

Create a set of commands to:

1. `subo docs build [dir] [--output]` recursively generate documentation with appropriate code snippets inserted at the directory `--output` (optional, otherwise [dir] is used as the output dir as well)
2. `subo docs test [dir]` recursively test the code snippets without documentation generation (although we do generate documentation to test the templates, but it's at subo's cache directory)

<details>
<summary> Details </summary>

This is done by using template actions and [testable examples](https://go.dev/blog/examples).
docs api uses the actions format `<!-- {{ Snippet "{go package}/[go function name]:[example name]" }} -->`, where package is the go package name and `example name` can be a package ("{go package}:[example name]") or a function example.
Note, only packages and functions with testable examples can be use, where the examples are the inserted code snippets. For example, `<!-- {{ Snippet "greetings/Hello:doThis"}} -->` tells us to insert the example code snippet `doThis` from the go test file with the package name `greetings` for the function `Hello`.
Note, you're allow to insert code snippets for [multi package and function examples](https://pkg.go.dev/testing#hdr-Examples).

</details>

# Usage

<details>
<summary>Template actions types</summary>

`{{ Snippet "greetings" }}` => all package examples are inserted from package `greetings`
`{{ Snippet "greetings:doNotDoThis" }}` => only package example `doNotDoThis` is  inserted from package `greeting`
`{{ Snippet "greetings/Hello" }}` => all function  `Hello` examples are inserted from package `greetings`
`{{ Snippet "greetings/Hello:doThis" }}` => only function  `Hello` example `doThis` is  inserted from package `greetings`

 </details>

# Errors

`subo doc` will error if you provide nonexistent examples for packages or functions

>`<!-- {{ Snippet "foo"}} -->`
```console
Error: failed to getUpdatedDocs: failed to Execute: template: docs/test/example.md:20:8: executing "docs/test/example.md" at <Snippet "foo">: error calling Snippet: <!-- {{ Snippet "foo" }} -->
Examples for the package `foo` do not exist
```

>`<!-- {{ Snippet "greetings/Bar" }} -->`
```console
Error: failed to getUpdatedDocs: failed to Execute: template: docs/test/example.md:20:8: executing "docs/test/example.md" at <Snippet "greetings/Bar">: error calling Snippet: <!-- {{ Snippet "greetings/Bar" }} -->
Examples for the function `Bar` in the package `greetings` do not exist
```

# Example

**How do we use these commands?**

First create a markdown file with template actions, where the actions are in the form of `<!-- {{ Snippet "{go package}/[go function name]:[example name]" }} -->`. Then create a Go file for the function and a test file with a testable example, where the test file matches the package name (both files can share the same package name).

For example, this is our current directory:
```console
(base) odc@mbp MyApp % ls
go.mod          greetings.go        greetings_test.go        test.md
```
<details>
<summary>Where greetings.go</summary>

> ```go
> package greetings
> import "fmt"
>
> func Hello(who string) {
> 	if who == "Boo!" {
> 		fmt.Printf("Ahhh!")
> 	} else {
> 		fmt.Printf("Hello, %s!\n", who)
> 	}
> }
> 
> ```

</details>

<details>
<summary>Where greetings_test.go</summary>

read over testable examples if unfamiliar, https://pkg.go.dev/testing#hdr-Examples

> ```go
> package greetings
> import "fmt"
>
>// This comment is associated with the ExampleHello_doNotDoThis function example.
>// This is showing you how not to use Hello
> func ExampleHello_doNotDoThis() {
>	Hello("Boo!")
>	// Output: Ahhh!
>}
>
> // This comment is associated with the ExampleHello_doThis function example.
> // This is showing you how to use Hello
> func ExampleHello_doThis() {
>	Hello("world")
>	// Output: Hello, world!
> }
> 
>
> // This comment is associated with the package example.
> // This is showing you how to use Hello
> func Example() {
>	Hello("Foo Bar")
>	// Output: Hello, Foo Bar!
> }
>
>// This comment is associated with the package example Example_doNotDoThis.
>// This is showing you how to use Hello
>func Example_doNotDoThis() {
>	Hello("boo!")
>	// Output: Hello, Bar Foo!
>}
> ```

</details>

<details>
<summary>Where test.md</summary>

> ```md
># This is a test
>
>## Existing Package Examples
>
><!-- {{ Snippet "greetings"}} -->
>
>## Package Example doNotDoThis
>
><!-- {{ Snippet "greetings:doNotDoThis"}} -->
>
>## Existing Function Examples
>
><!-- {{ Snippet "greetings/Hello" }} -->
>
>
>## Existing Function Example doThis
>
><!-- {{ Snippet "greetings/Hello:doThis" }} -->
> ```

</details>

Running the docs build command `subo docs build . --output ../temp` will  create a file `../temp/test.md`

<details>
<summary>Where temp/test.md</summary>

># This is a test
>
>## Existing Package Examples
>
><!-- {{ Snippet "greetings" }} -->
><!-- DO NOT REMOVE: START -->
>
>```go
>Hello("Foo Bar")
>
>```
>
>```go
>Hello("boo!")
>
>```
><!-- DO NOT REMOVE: END -->
>## Package Example doNotDoThis
>
><!-- {{ Snippet "greetings:doNotDoThis" }} -->
><!-- DO NOT REMOVE: START -->
>
>```go
>Hello("boo!")
>
>```
><!-- DO NOT REMOVE: END -->
>## Existing Function Examples
>
><!-- {{ Snippet "greetings/Hello" }} -->
><!-- DO NOT REMOVE: START -->
>
>```go
>Hello("Boo!")
>
>```
>
>```go
>Hello("world")
>
>```
><!-- DO NOT REMOVE: END -->
>## Existing Function Example doThis
>
><!-- {{ Snippet "greetings/Hello:doThis" }} -->
><!-- DO NOT REMOVE: START -->
>
>```go
>Hello("world")
>
>```
><!-- DO NOT REMOVE: END -->
>

</details>

<details>
<summary>Normalized template actions</summary>

`subo docs` will remove the old `<-- Do NOT REMOVE END/START -->` pairs every time it is ran and insert the new doc pairs. 
`subo docs` also normalizes the template action `<!-- {{ Snippet "go" "greetings" "Hello" }} -->`. For examples, the following variations are allow:

```md
<!--     {{ Snippet "go" "greetings" "Hello" }}      -->

<!--

   {{ Snippet "go" "greetings" "Hello" }}
          -->
```
Would be normalized to:
`<!--     {{ Snippet "go" "greetings" "Hello" }}      -->`

</details>

# Test

To test this you can run `subo doc build` at subo/docs/test/, `subo doc build subo/docs/test`

jagger27 commented:
This is neat, but I'm not sure I get what it has to do with #139.

denopink commented:
> This is neat, but I'm not sure I get what it has to do with #139.

omg, you are so right @jagger27 ! That was a typo... it should have been #22 not #139 😆

jagger27 commented on subo/command/docs.go:
```diff
@@ -0,0 +1,171 @@
+		return nil, errors.Wrap(err, "failed to Stat")
+	}
+
+	// go build
+	goBuildArgs := []string{"build"}
+	if goBuildFlags != nil {
+		// go build [-o output] [build flags]
+		goBuildArgs = append(goBuildArgs, goBuildFlags...)
+	}
+	// go build [-o output] [build flags] [packages]
+	goBuildArgs = append(goBuildArgs, []string{dir}...)
+
+	if output, err := exec.Command("go", goBuildArgs...).CombinedOutput(); err != nil {
+		return nil, errors.Wrap(err, fmt.Sprintf("failed to build code due to:\n%s", output))
+	}
```
Go already has a nice idiomatic way of testing code snippets as part of the `go doc` tool along with `go test`. Doing it this way, we can get Examples sections for free in `go docs` as well.

I haven't looked too far into it, but I'm pretty sure we can use parts of the `go doc` codebase to extract snippets of code and use that instead of rolling our own file system crawling stuff.

Let's talk about this more! I have a few ideas.

jagger27 replied:
See: https://go.dev/blog/examples

denopink replied:
@jagger27 I went through both `go doc` and `godoc` code base and only `godocs` actually extracts the testable examples and formats everything (no static doc api either) nicely like in that [shared example](https://go.dev/blog/examples).

The good news is that I converted `subo docs` to leverage the same code extraction as `godocs`. Unfortunately, we can't easily leverage godocs formatting since those apis are baked into godocs server logic (no static doc apis).

Our formatting code is actually pretty basic since we only want the testable examples and not everything that `godoc` formatting offers, so it's pretty light. So I'm not sure there's much benefit in adapting godocs formatting
since most of the formatting code is private and baked into godoc server logic.

Let me know if you like this formatting or not. I updated the main example up above.

denopink commented:
@jagger27 I went through both `go doc` and `godoc` code base and only `godocs` actually extracts the testable examples and formats everything (no static doc api either) nicely like in that [shared example](https://go.dev/blog/examples).

The good news is that I converted `subo docs` to leverage the same code extraction as `godocs`. Unfortunately, we can't easily leverage godocs formatting since those apis are baked into godocs server logic (no static doc apis).

Our formatting code is actually pretty basic since we only want the testable examples and not everything that `godoc` formatting offers, so it's pretty light. So I'm not sure there's much benefit in adapting godocs formatting
since most of the formatting code is private and baked into godoc server logic.

Let me know if you like this formatting or not. I updated the main example up above.

jagger27 reviewed (commented):
I've had some time to think about this and I have a few new ideas :)

Let's not bother with function comments and `// Output: foobar` bits. The snippet should only inject the code sample itself. My reasoning here is that the comments might not make sense in the context of that specific place in the docs, especially if the `## Header Levels` don't match up. Simpler is better here I think.

The other thing I'd like to see is slightly simpler syntax, which I think should be possible.

Here's my suggestion:
```markdown
<!-- {{ Snippet "go" "greetings" "Hello" }} -->
```
Which compiles to:
~~~markdown
<!-- {{ Snippet "go" "greetings" "Hello" }} -->
```go
Hello("Boo!")
```
<!-- END Snippet -->
~~~

I think having the `<!-- DO NOT REMOVE: START -->` block is redundant because we can infer the start of the snippet from the template chunk itself. If there is a technical need to differentiate between a fresh snippet and one that was already inserted, maybe the snippet opener could be replaced with `ExistingSnippet` or something? It would be better not to have to do that, though:
~~~markdown
<!-- {{ ExistingSnippet "go" "greetings" "Hello" }} -->
~~~

Another little aside: I think it should be possible to parse without an explicit terminal comment by looking for the last \`\`\` since it will always be there, but it might be clearer where things start and end and more obvious that the code wasn't inserted by hand if there is a comment present.

denopink commented:
@jagger27 I agree, having the comment(s) makes it clearer where things end and start.

atm it's easier to keep the two comments (start/end) since it makes for a simple regex to remove old docs.

I also agree with just keeping the code and not the extra docs and headers.

denopink commented:
@jagger27 the current update will error if you provide nonexistent examples for packages or functions

>`<!-- {{ Snippet "foo"}} -->`
```console
Error: failed to getUpdatedDocs: failed to Execute: template: docs/test/example.md:20:8: executing "docs/test/example.md" at <Snippet "foo">: error calling Snippet: <!-- {{ Snippet "foo" }} -->
Examples for the package `foo` do not exist
```

>`<!-- {{ Snippet "greetings/Bar" }} -->`
```console
Error: failed to getUpdatedDocs: failed to Execute: template: docs/test/example.md:20:8: executing "docs/test/example.md" at <Snippet "greetings/Bar">: error calling Snippet: <!-- {{ Snippet "greetings/Bar" }} -->
Examples for the function `Bar` in the package `greetings` do not exist
```

We also removed `---` from the snippets and the extra newline in the end.

We simplified the api to the following:
Simplify `doc` text template action to the following:

`{{ Snippet "greetings" }}` => all package examples are inserted from package `greetings`
`{{ Snippet "greetings:doNotDoThis" }}` => only package example `doNotDoThis` is  inserted from package `greeting`
{{ Snippet "greetings/Hello" }} => all function  `Hello` examples are inserted from package `greetings`
{{ Snippet "greetings/Hello:doThis" }} => only function  `Hello` example `doThis` is  inserted from package `greetings`

jagger27 commented on subo/command/docs.go:
````diff
@@ -0,0 +1,484 @@
+			&exampleData{
+				Suffix: example.Suffix,
+				Doc:    example.Doc,
+				Code:   fmt.Sprintf("```go\n%s\n```", buffer.String()),
````
```suggestion
				Code:   fmt.Sprintf("```go\n%s```", buffer.String()),
```

Ah-ha! This is the newline I was talking about.

denopink replied:
> lol, whops I forgot about that newline!

Looks like I stashed this changed locally some 🙃

jagger27 reviewed (approved):
Besides that one little newline thing, this LGTM.

Final steps before merging:
- rebase on main
- squash commits and use Conventional Commits (like you did for the PR title).

denopink commented:
> Besides that one little newline thing, this LGTM.
> 
> Final steps before merging:
> 
>     * rebase on main
> 
>     * squash commits and use Conventional Commits (like you did for the PR title).

Alright, we rebased and squashed.

javorszky reviewed (approved):
looks good, works as expected :) :shipit:
